### DISCUSSION TITLE
Can firmware be saved locally when upgrading?

### DISCUSSION THREAD

**Discussion by cihun** [fwupd/fwupd#5592] [Q&A]
I'm working on a firmware switch thing that may flash different firmware when the SIM card is switched.

If you do not have the correct firmware for flash, you may not be able to connect to the network.

Therefore, I wonder whether the firmware package can be saved and downloaded when the flash firmware is successful? When the SIM card is switched, the firmware corresponding to the flash can be downloaded.

**Comment by hughsie** [comment]
Yes, we could do that. We already save a backup firmware for some hardware (i.e. save the old version before writing the new version) although I'm not quite sure how your proposed functionality can be added to the existing API. Counter proposal -- could all the different firmware versions be saved into one cab archive, and we save *that* somewhere persistent? e.g. one thing to download, one thing to store. That would be a lot easier.

**Reply by cihun** [reply]
> Yes, we could do that.

Is there a storage interface now? If yes, can you tell me the documentation?

> could all the different firmware versions be saved into one cab archive, and we save that somewhere persistent?

Yes, we do store multiple different versions of firmware in a cab file.

Thanks.

**Reply by hughsie** [reply]
> Is there a storage interface now? If yes, can you tell me the documentation?

There's FWUPD_DEVICE_FLAG_BACKUP_BEFORE_INSTALL used in https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-plugin.c#L759

It's likely we'd want something like FU_DEVICE_INTERNAL_FLAG_RETAIN_FIRMWARE_ARCHIVE which does the same thing in fu-engine.c -- if that sounds about correct let me know and I can do a prototype.

**Reply by cihun** [reply]
I am upgrading the modem. Different carriers may require different versions of firmware, so when users insert different SIM cards, we need to determine the correct firmware to flash when needed. 

I intend to put all the firmware in the cab package, which contains four types of firmware (modem, ap, oem, op), each of which may have multiple versions of firmware. 

So when the user updates the cab package, I want to be able to save the cab package locally so that when the user switches SIM cards, they can do offline flash without a network.

Thanks.

**Reply by hughsie** [reply]
https://github.com/fwupd/fwupd/pull/5598 might be interesting for you.

**Comment by cihun** [comment]
Hi hughsie,

I want flash firmware in the background, which interface should I use? 
I am making the SIM firmware switch for the modem, different SIM cards may map different firmware. 
I can use the modem-manager plug-in to achieve this state when the user inserts or replaces a SIM card. 
But at this point, I don't know how to trigger the flash to get the device firmware to switch to the correct state.

Thanks.

**Reply by hughsie** [reply]
> I want flash firmware in the background

`fwupdmgr reinstall MODEM_GUID` should do the right thing if the firmware has been saved locally.

**Reply by cihun** [reply]
I want to add the function of SIM card detection to the modem-manager plug-in of fwupd. If a SIM card is inserted, It detects the firmware version corresponding to the SIM card and determine if flash is needed.

The process will be automatic.

Thanks.

**Reply by hughsie** [reply]
I don't think a plugin should install firmware automatically without user interaction.

**Reply by cihun** [reply]
we believe by insert a new SIM card, it will completely showing the motivation of end user to use the new SIM card. So fw should be will prepared at this moment.

Do all plug-ins in fwupd not install automatically?

Can all installations be called only by fwupdmgr or software?

Thanks.

**Reply by hughsie** [reply]
> Do all plug-ins in fwupd not install automatically

No -- plugins just enumerate, provide details, and perform updates. They can't really dictate what the daemon does.

**Reply by cihun** [reply]
Thanks, I think I might need to write another daemon to trigger flash.

**Reply by hughsie** [reply]
There certainly needs to be something that's listening for the SIM card to be re-inserted and to trigger fwupd into doing something -- but I think it's a pretty poor user experience generally to have to flash new firmware just to change a SIM. It's not something you have to do on a mobile phone, right?

**Reply by cihun** [reply]
This should not lead to a bad user experience, we have the firmware divided into 4 or 5 parts, do not need to be all flash, most of the time users switch SIM cards only need a flash op, the size of the op is less than 3MB. 

We also have this feature in ChromeOS and Windows, and now we need to implement the same feature in Ubuntu.

**Comment by cihun** [comment]
Modems in flash are required to flash four or five firmware packages, which may result in different combinations due to changes in the SIM card. I want to flash all the firmware correctly at once. Do you have any suggestions for this?

Thanks.

**Comment by cihun** [comment]
I would like to implement the flash function of the module in fwupd, which can be detected by the modem-manager plug-in and create an MmDevice object, and flash using fastboot.

Since there are several different firmware versions, I created several subdevices to track the version numbers of these firmware. I also split the firmware in the cab package into several parts, each with a separate metainfo.xml file, and added restrictions to the requirements.

Because the firmware belongs to the same device. As a result, the device needs to perform detach and attach each time it flashes one firmware, which causes the module to fail to flash the second firmware after the first firmware flash.

Can you add a feature that allows fwupd to flash releases with the same device together between detach and attach?

Or do you have any other suggestions?

Thanks.

**Reply by hughsie** [reply]
>  I created several subdevices to track the version numbers of these firmware

Subdevices in fwupd?

> the device needs to perform detach and attach each time it flashes one firmware

There's also ->prepare() and ->cleanup(), although I am a bit confused why the 2nd firmware flash fails. Is the modem not coming back online after the 1st flash?

**Reply by cihun** [reply]
> Subdevices in fwupd?

Sorry, is a child device.

>  Is the modem not coming back online after the 1st flash?

I have found the cause of the problem. plugin's detach is executed first when executing device's detach. In modem-manager's plugin's detach, another FuMmDevice object is created. This object does not have omodem so it cannot enter the bootloader state, resulting in no flash.

I also noticed that, in the modem - manager detach, can inhibit modemmanager, and in the attach, will uninbit modemmanager. But when the flash module using the fastboot plugin, therefore, The attach of the modem-manager plugin is not executed. Procedure As a result, the modem-manager could not find the module after the flash finished. 

This also causes fwupd to be unable to find the module.

Can we set a switch that triggers flash? This switch is turned off by default and the user can decide whether to turn it on or not.

Thanks.

**Reply by hughsie** [reply]
> This also causes fwupd to be unable to find the module.

Can you do a patch that fixes both issues? I don't have any suitable hardware here for testing.

> Can we set a switch that triggers flash?

You mean something in the config file? I don't think that's appropriate. I think fwupd should just do the action, not impose policy.

**Reply by superm1** [reply]
Conceptually I think what you are looking for is modem manager to behave as a fwupd client and request the daemon to do the flash. Then the policy is decided by that client.

**Comment by cihun** [comment]
> Can you do a patch that fixes both issues? 

Yes, but I'm not sure if this is a  issues because the modem-manager plugin is also used by other people. 

> You mean something in the config file? 

Yes. 

Our firmware requires certification, which is based on operator, so it is possible that some operators of the new version may pass the certification. 

So we need to implement fwswitch to switch firmware between users using different SIM cards. 

If the user thinks flash should not be executed when switching SIM cards, he can choose not to turn it on. 

If the user turns this on, we assume he is ready for flash when switching SIM cards. 

> Conceptually I think what you are looking for is modem manager to behave as a fwupd client and request the daemon to do the flash. Then the policy is decided by that client.

I'll think about the proposal.

**Comment by cihun** [comment]
Does fwupd only provide interfaces as apis?

Do other device upgrades have to be done by external programs (e.g. fwupdmgr) using fwupd's interface?

Thanks.

**Reply by superm1** [reply]
> Does fwupd only provide interfaces as apis?
> Do other device upgrades have to be done by external programs (e.g. fwupdmgr) using fwupd's interface?

Yeah, that's right.  Nothing in the daemon happens automatically for flashing.  That's why I was suggesting if you want to cause an automatic flash of something the policy should be controlled by a fwupd client.
